File: measurement-generator/src/main/java/pl/polsl/hdised/measurementgenerator/measurement/MeasurementCreator.java
package pl.polsl.hdised.measurementgenerator.measurement;

import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

import java.util.Calendar;
import java.util.Random;
import java.util.concurrent.atomic.AtomicBoolean;

@Component
public class MeasurementCreator {

    private AtomicBoolean produce;

    // IF WE WANT TO WORK WITH DIFFERENT CLASS THAN STRING WE NEED TO CHANGE THIS SECOND ARGUMENT
    private KafkaTemplate<String, MeasurementDto> measurementKafkaTemplate;


    public MeasurementCreator(KafkaTemplate<String, MeasurementDto> measurementKafkaTemplate) {
        this.measurementKafkaTemplate = measurementKafkaTemplate;
        this.produce = new AtomicBoolean(true);
    }

    public synchronized void setProduce(boolean produce) {
        this.produce.set(produce);
    }

    public void StartProducingMeasurementsToDatabase(String topic) throws InterruptedException {
        while (true) {
            if (this.produce.get()) {
                MeasurementDto measurementDto = createMeasurement();
                System.out.println(measurementDto.getCityName());
                System.out.println(measurementDto.getDeviceId());
                System.out.println(measurementDto.getDate());
                System.out.println(measurementDto.getUnit());
                System.out.println(measurementDto.getTemperature());
                System.out.println("----------------------------");
                measurementKafkaTemplate.send(topic, measurementDto);
                Thread.sleep(500);
            } else {
                break;
            }
        }
    }

    private MeasurementDto createMeasurement(){
        String city;
        switch(new Random().nextInt(0, 3)){
            case 0 -> city = "Gliwice";
            case 1 -> city = "Katowice";
            case 2 -> city = "Warszawa";
            default -> city = "Wroclaw";
        }
        String deviceClass;
        switch(new Random().nextInt(0, 3)){
            case 0 -> deviceClass = "dev01";
            case 1 -> deviceClass = "dev02";
            case 2 -> deviceClass = "dev03";
            default -> deviceClass = "dev04";
        }
        Float temperature = new Random().nextFloat(-30.0f, 30.0f);
        return new MeasurementDto(city, Calendar.getInstance().getTime(), deviceClass, temperature, "Celsius");

    }

}

File: golang-measurement-generator/measurementCreator.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math/rand"
	"time"

	kafka "github.com/segmentio/kafka-go"
)

type MeasurementCreator struct {
	Produce bool
	Writer  *kafka.Writer
}

func NewMeasurementCreator(writer *kafka.Writer) *MeasurementCreator {
	return &MeasurementCreator{
		Produce: true,
		Writer:  writer,
	}
}

func (mc *MeasurementCreator) SetProduce(produce bool) {
	mc.Produce = produce
}

func (mc *MeasurementCreator) StartProducingMeasurementsToDatabase(topic string) {
	mc.Writer.Topic = topic
	for mc.Produce {
		measurementDto := mc.createMeasurement()
		fmt.Println(measurementDto.CityName)
		fmt.Println(measurementDto.DeviceId)
		fmt.Println(measurementDto.Date)
		fmt.Println(measurementDto.Unit)
		fmt.Println(measurementDto.Temperature)
		fmt.Println("----------------------------")

		jsonData, err := json.Marshal(measurementDto)
		if err != nil {
			fmt.Println(err)
			return
		}
		measurementByte := []byte(jsonData)
		msg := kafka.Message{
			Value: []byte(measurementByte),
		}
		err = mc.Writer.WriteMessages(context.Background(), msg)
		if err != nil {
			fmt.Println(err)
		}
		time.Sleep(500 * time.Millisecond)
	}
}

func (mc *MeasurementCreator) createMeasurement() *MeasurementDto {
	var city string
	switch rand.Intn(3) {
	case 0:
		city = "Gliwice"
	case 1:
		city = "Katowice"
	case 2:
		city = "Warszawa"
	default:
		city = "Wroclaw"
	}
	var deviceClass string
	switch rand.Intn(3) {
	case 0:
		deviceClass = "dev01"
	case 1:
		deviceClass = "dev02"
	case 2:
		deviceClass = "dev03"
	default:
		deviceClass = "dev04"
	}
	temperature := rand.Float64()*60 - 30
	return NewMeasurementDto(city, deviceClass, time.Now(), "Celsius", temperature)
}
